lib/session.ex

```diff
@@ -0,0 +1,195 @@
+      {:ok, %{"id" => session_id}} ->
+        start_link(session_id, connection)
+
+      {:error, reason} ->
+        {:error, reason}
+    end
+  end
+
+  @spec start_link(session_id_t(), connection_t()) :: GenServer.server()
+  def start_link(session_id, connection) do
+    GenServer.start_link(__MODULE__, {session_id, connection}, [])
+  end
+
+  @doc """
+  Synchronously attaches a plugin to the session on the gateway.
+
+  ## Arguments
+
+  * `session` - a PID of the `Janus.Session` process,
+  * `plugin` - a string containing valid gateway's plugin name,
+  * `timeout` - a timeout for the call.
+
+  ## Return values
+
+  On success it returns `{:ok, plugin_handle_id}` where `plugin_handle_id` is
+  a positive integer which is used as internal plugin handle identifier by the
+  gateway.
+
+  On error it returns `{:error, reason}`.
+
+  The reason might be:
+
+  * `{:gateway, code, info}` - it means that the call itself succeded but the
+    gateway returned an error of the given code and info,
+  * other - some serious error happened.
+  """
+  @spec session_attach(Janus.Session.t(), plugin_t(), timeout_t()) ::
+          {:ok, plugin_handle_id} | {:error, any}
+  def session_attach(session, plugin, timeout \\ @default_timeout) do
+    case __MODULE__.execute_request(
+           session,
+           %{janus: :attach, plugin: plugin},
+           timeout
+         ) do
+      {:ok, %{"id" => plugin_handle_id}} ->
+        {:ok, plugin_handle_id}
+
+      {:error, reason} ->
+        {:error, reason}
+    end
+  end
+
+  @doc """
+  Adds `session_id` to given message and sends it through connection stored by given session.
+
+  ## Arguments
+
+  * `session` - a PID of `Janus.Session` process.
+  * `message` - map containing message request.
+  * `timeout` - timeout passed to `Janus.Connection.call/3`.
+
+  ## Return values
+
+  Returns response same as `Janus.Connection.call/3`.
+  """
+  @spec execute_request(Janus.Session.t(), message_t(), timeout_t()) :: {:ok, any} | {:error, any}
+  def execute_request(session, message, timeout \\ @default_timeout) do
+    GenServer.call(session, {:execute_message, message, timeout})
+  end
+
+  @doc """
+  Replaces current connection with a new one.
+
+  ## Arguments
+
+  * `session` - a PID of `Janus.Session` process.
+  * `connection` - a PID of `Janus.Connection` process.
+
+  ## Return values
+
+  Returns :ok atom.
+  """
+  @spec update_connection(Janus.Session.t(), connection_t()) :: :ok
+  def update_connection(session, connection) do
+    GenServer.cast(session, {:new_connection, connection})
+  end
+
+  # callbacks
+
+  @impl true
+  def init({session_id, connection}) do
+    state = %{
+      session_id: session_id,
+      connection: connection
+    }
+
+    keep_alive = Janus.Connection.get_transport_module(connection).needs_keep_alive?()
+
+    case keep_alive do
+      {true, keep_alive_timeout} ->
+        Process.send_after(self(), :keep_alive, keep_alive_timeout)
```

Comment:
I would advice making call sooner then keep_alive_timeout. Process. send_after has no guarantee of being on time.

Reply:
My goal was to force every transport to set proper keep alive timeout, by proper I mean the interval when the message should be sent instead of indicating when janus will drop session. So for instance in WS transport this timeout should be set to something like 30 seconds where Janus will drop session in 60 seconds.